crates/bitwarden-crypto/src/traits/encryptable.rs

```diff
@@ -0,0 +1,250 @@
+use crate::{store::KeyStoreContext, AsymmetricEncString, CryptoError, EncString, KeyId, KeyIds};
+
+/// An encryption operation that takes the input value and encrypts it into the output value.
+/// Implementations should generally consist of calling [Encryptable::encrypt] for all the fields of
+/// the type.
+pub trait Encryptable<Ids: KeyIds, Key: KeyId, Output> {
+    fn encrypt(&self, ctx: &mut KeyStoreContext<Ids>, key: Key) -> Result<Output, CryptoError>;
+}
+
+impl<Ids: KeyIds> Encryptable<Ids, Ids::Symmetric, EncString> for &[u8] {
+    fn encrypt(
+        &self,
+        ctx: &mut KeyStoreContext<Ids>,
+        key: Ids::Symmetric,
+    ) -> Result<EncString, CryptoError> {
+        ctx.encrypt_data_with_symmetric_key(key, self)
+    }
+}
+
+impl<Ids: KeyIds> Encryptable<Ids, Ids::Asymmetric, AsymmetricEncString> for &[u8] {
+    fn encrypt(
+        &self,
+        ctx: &mut KeyStoreContext<Ids>,
+        key: Ids::Asymmetric,
+    ) -> Result<AsymmetricEncString, CryptoError> {
+        ctx.encrypt_data_with_asymmetric_key(key, self)
+    }
+}
+
+impl<Ids: KeyIds> Encryptable<Ids, Ids::Symmetric, EncString> for Vec<u8> {
+    fn encrypt(
+        &self,
+        ctx: &mut KeyStoreContext<Ids>,
+        key: Ids::Symmetric,
+    ) -> Result<EncString, CryptoError> {
+        ctx.encrypt_data_with_symmetric_key(key, self)
+    }
+}
+
+impl<Ids: KeyIds> Encryptable<Ids, Ids::Asymmetric, AsymmetricEncString> for Vec<u8> {
+    fn encrypt(
+        &self,
+        ctx: &mut KeyStoreContext<Ids>,
+        key: Ids::Asymmetric,
+    ) -> Result<AsymmetricEncString, CryptoError> {
+        ctx.encrypt_data_with_asymmetric_key(key, self)
+    }
+}
+
+impl<Ids: KeyIds> Encryptable<Ids, Ids::Symmetric, EncString> for &str {
+    fn encrypt(
+        &self,
+        ctx: &mut KeyStoreContext<Ids>,
+        key: Ids::Symmetric,
+    ) -> Result<EncString, CryptoError> {
+        self.as_bytes().encrypt(ctx, key)
+    }
+}
+
+impl<Ids: KeyIds> Encryptable<Ids, Ids::Asymmetric, AsymmetricEncString> for &str {
+    fn encrypt(
+        &self,
+        ctx: &mut KeyStoreContext<Ids>,
+        key: Ids::Asymmetric,
+    ) -> Result<AsymmetricEncString, CryptoError> {
+        self.as_bytes().encrypt(ctx, key)
+    }
+}
```

Comment:
I believe we can generalize this code slightly.

```suggestion
impl<Ids, K, E> Encryptable<Ids, K, E> for &str
where
    Ids: KeyIds,
    K: KeyId,
    E: EncStringType,
    [u8]: Encryptable<Ids, K, E>,
{
    fn encrypt(&self, ctx: &mut KeyStoreContext<Ids>, key: K) -> Result<E, CryptoError> {
        self.as_bytes().encrypt(ctx, key)
    }
}
```

And define a common trait for `EncString`.
```rust
pub trait EncStringType {}
impl EncStringType for EncString {}
impl EncStringType for AsymmetricEncString {}
```

Reply:
This is something Matt also brought up when trying to do authenticated data. I think the best path forward would be to separate encryptable/decrytable into encryptable/decryptable and encodable/decodable, something like this.

```rust
impl <T: Encodable<Vec<u8>> Encryptable<EncString> for T
impl <T: Encodable<Vec<u8>> Encryptable< AsymmetricEncString> for T

impl Encodable<Vec<u8>> for &str
impl Encodable<Vec<u8>> for String
```